src/Microsoft.ComponentDetection.Contracts/FileComponentDetectorWithCleanup.cs

```diff
@@ -0,0 +1,154 @@
+        this.DirectoryUtilityService = directoryUtilityService;
+    }
+
+    protected IFileUtilityService FileUtilityService { get; private set; }
+
+    protected IDirectoryUtilityService DirectoryUtilityService { get; private set; }
+
+    /// <summary>
+    /// Patterns of files and folders that should be cleaned up after the detector has run, if they were created by the detector.
+    /// </summary>
+    protected virtual IList<string> CleanupPatterns { get; set; }
+
+    /// <summary>
+    /// Takes a process and wraps it in a cleanup operation that will delete any files or
+    /// directories created during the process that match the clean up patterns.
+    /// </summary>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+    protected async Task WithCleanupAsync(
+        Func<ProcessRequest, IDictionary<string, string>, CancellationToken, Task> process,
+        ProcessRequest processRequest,
+        IDictionary<string, string> detectorArgs,
+        bool cleanupCreatedFiles,
+        CancellationToken cancellationToken = default)
+    {
+        if (process == null)
+        {
+            throw new ArgumentNullException(nameof(process));
+        }
+
+        // If there are no cleanup patterns, or the relevant file does not have a valid directory, run the process without even
+        // determining the files that exist as there will be no subsequent cleanup process.
+        if (this.FileUtilityService == null
+            || this.DirectoryUtilityService == null
+            || !this.TryGetCleanupFileDirectory(processRequest, out var fileParentDirectory))
+        {
+            await process(processRequest, detectorArgs, cancellationToken).ConfigureAwait(false);
+            return;
+        }
+
+        // Get the files and directories that match the cleanup pattern and exist before the process runs.
+        var (preExistingFiles, preExistingDirs) = this.DirectoryUtilityService.GetFilesAndDirectories(fileParentDirectory, this.CleanupPatterns, DefaultCleanDepth);
+        try
+        {
+            await process(processRequest, detectorArgs, cancellationToken).ConfigureAwait(false);
+        }
+        finally
+        {
+            // Ensure that only one cleanup process is running at a time, helping to prevent conflicts
+            await this.cleanupSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            try
+            {
+                // Clean up any new files or directories created during the scan that match the clean up patterns.
+                // If the cleanupCreatedFiles flag is set to false, this will be a dry run and will just log the files that it would clean.
+                var dryRun = !cleanupCreatedFiles;
+                var dryRunStr = dryRun ? "[DRYRUN] " : string.Empty;
+                var (newFiles, newDirs) = this.DirectoryUtilityService.GetFilesAndDirectories(fileParentDirectory, this.CleanupPatterns, DefaultCleanDepth);
```

Comment:
nit: `newFile` and `newDir` names give an impression of these being the newly created file. can we change it to something different for e.g. `latestFiles` `latestDirs`